notifier/stomp/directdeliverer.go

```diff
@@ -0,0 +1,99 @@
+package stomp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	gostomp "github.com/go-stomp/stomp"
+	"github.com/google/uuid"
+	clairerror "github.com/quay/clair/v4/clair-error"
+	"github.com/quay/clair/v4/notifier"
+)
+
+// Deliverer is a STOMP deliverer which publishes a notifier.Callback to the
+// the broker.
+type DirectDeliverer struct {
+	conf Config
+	n    []notifier.Notification
+	fo   *failOver
+}
+
+func NewDirectDeliverer(conf Config) (*DirectDeliverer, error) {
+	var c Config
+	var err error
+	if c, err = conf.Validate(); err != nil {
+		return nil, err
+	}
+	fo := &failOver{
+		Config: c,
+	}
+	return &DirectDeliverer{
+		conf: c,
+		n:    []notifier.Notification{},
+		fo:   fo,
+	}, nil
+}
+
+func (d *DirectDeliverer) Name() string {
+	return fmt.Sprintf("stomp-direct-%s", d.conf.Destination)
+}
+
+// Notifications will copy the provided notifications into a buffer for STOMP
+// delivery.
+func (d *DirectDeliverer) Notifications(ctx context.Context, n []notifier.Notification) error {
+	// if we can reslice instead of allocate do so.
+	if len(n) <= len(d.n) {
+		d.n = d.n[:len(n)]
+		copy(d.n, n)
+		return nil
+	}
+	tmp := make([]notifier.Notification, len(n), len(n))
+	copy(tmp, n)
+	d.n = tmp
+	return nil
+}
+
+func (d *DirectDeliverer) Deliver(ctx context.Context, nID uuid.UUID) error {
+	conn, err := d.fo.Connection(ctx)
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{err}
+	}
+	defer conn.Disconnect()
+
+	tx, err := conn.BeginWithError()
+	if err != nil {
+		return &clairerror.ErrDeliveryFailed{err}
+	}
+
+	// block loop publishing smaller blocks of max(rollup) length via reslicing.
+	var rollup int = d.conf.Rollup
+	if rollup == 0 {
+		rollup++
+	}
+	for bs, be := 0, rollup; bs < len(d.n); bs, be = be, be+rollup {
+		// if block-end exceeds array bounds, slice block underflow.
+		// next block-start will cause loop to exit.
+		if be > len(d.n) {
+			be = len(d.n)
+		}
+
+		n := d.n[bs:be]
+		b, err := json.Marshal(&n)
```

Comment:
This will generate a lot of separate slices. I think it will be easier on the gc to reuse a bytes.Buffer and json.Encoder for each iteration of the loop, e.g.:

```
buf := &bytes.Buffer{}
enc := json.NewEncoder(buf)
for ... {
    ...
    buf.Reset()
    if err := enc.Encode(&n); err != nil {
        ...
    }
    ...
    tx.Send(..., buf.Bytes(), ...)
}
```

This would result in just two big buffers getting used, rather than 1 + number of iterations.

Its a real shame there's not an API for the stdlib json that doesn't involve buffering the whole data structure internally.

Reply:
fixed